Fix DistanceWithin algorithm with PostgreSQL provider

strk opened the pull request:
This PR is to work toward fixing the bug reported in #45352

It starts with a testcase which I'm hoping will fail (can't verify myself due to #45380)

strk committed:
Add (intended-to-fail) test for https://github.com/qgis/QGIS/issues/45352

strk commented:
Unfortunately the test is NOT failing so it needs more action (maybe different configurations, should check out exactly how the request is constructed by QgsDistanceWithinAlgorithm

strk commented:
The QgsDistanceWithinAlgorithm is constructing the request like this:

> QgsFeatureRequest().setDistanceWithin( f.geometry(), currentDistance ).setNoAttributes().setDestinationCrs( targetSource->sourceCrs(), context.transformContext() );

Now I feel like a test specific to the QgsDistanceWithinAlgorithm would be more appropriate, as it could be a misuse of some kind ?

strk committed:
Add human-readable errors upon not meeting expected test preconditions

nyalldawson commented:
@strk 
> Now I feel like a test specific to the QgsDistanceWithinAlgorithm would be more appropriate, as it could be a misuse of some kind ?

That sounds fine to me -- it would belong in testqgsprocessingalgs.cpp

strk committed:
Add equality/inequality operators for QgsFeatureRequest::QgsSimplifyMethod

Includes unit test

The equality operator is useful for unit tests, the inequality
operator was added for completeness as requested by Nyall
in https://github.com/qgis/QGIS/pull/45384#discussion_r734174883

strk committed:
Add equality/inequality operators for QgsFeatureRequest::OrderBy

Equality useful for unit tests, inequality for completeness

strk committed:
Add unit test for QgsFeatureRequest

Tests the default constructor, the copy constructor and
the assignment operator.

strk committed:
Add test for ExtractWithinDistance with PostgreSQL layer inputs

This is a test for the bug reported in #45352

strk commented:
Now the test is ok (fails on my machine and is confirmed to PASS if I drop the `setLimit(1)` line in processIteratingOverTarget.
I will let GitHub workflow confirm the failure and then push a commit which changes QgsFeatureRequest::setWithinDistance so to set the mFilterRect member to NULL instead of the buffer around the reference geometry, as I think such operation should be done by providers if that's the way they want to implement the `within distance` filter.

strk commented:
Bug confirmed by CI: https://github.com/qgis/QGIS/runs/3966231575?check_suite_focus=true#step:13:994

strk commented:
The suggested fix of setting mFilterRect to NULL is not good because the ::accept member of QgsFeatureRequest is *using* that rectangle, and even expecting it to be in the correct CRS, although it's delegating that to the caller, a lot of confusion

strk committed:
Simplify assert

strk committed:
Add QgsAbstractFeatureIterator::updateRequestToSourceCrs method

The method superceeds filterRectToSourceCrs, deprecating it for
shortage of functionality now that we support multiple spatial
filters.

Fix #45352

nyalldawson commented on tests/src/core/testqgsfeaturerequest.cpp:
```diff
@@ -0,0 +1,153 @@
+
+}
+
+void TestQgsFeatureRequest::testDefaultConstructed( const QgsFeatureRequest &f )
```
Note that there's already a bunch of tests in test_qgsfeaturerequest.py -- there may be duplicates here now

nyalldawson replied:
It's not a blocker, but in general it's nicer to keep the tests together so that it avoids duplication of effort

nyalldawson commented on tests/src/core/testqgsfeaturerequest.cpp:
```diff
@@ -0,0 +1,153 @@
+  // is NOT FOUND by github CI, see
+  // - https://github.com/qgis/QGIS/pull/45384/checks?check_run_id=3936879569
+  // - https://cdash.orfeo-toolbox.org/viewBuildError.php?buildid=87673
+  //f2.setSubsetOfAttributes( {"a", "b"} );
```
setSubsetOfAttributes needs a list of integers, not strings

strk commented:
Please don't let this close automatically. Fixes an actual bug. Contains unit tests showing the bug and proving the fix.

nyalldawson commented on src/core/qgsfeatureiterator.cpp:
```diff
@@ -120,6 +121,37 @@ void QgsAbstractFeatureIterator::geometryToDestinationCrs( QgsFeature &feature,
+    case Qgis::SpatialFilterType::DistanceWithin:
+    {
+      QgsGeometry geom = request.referenceGeometry();
+      QgsRectangle bbox = geom.boundingBox();
+      double x0 = bbox.xMaximum();
+      double y0 = bbox.yMaximum();
+      double dist = request.distanceWithin();
+      QgsPoint p0( x0, y0 );
+      QgsPoint p1( x0 + dist, y0 );
+      QgsLineString newDistLine( p0, p1 );
+      newDistLine.transform( transform, Qgis::TransformDirection::Reverse );
+      dist = newDistLine.length();
```
I don't think this is a safe approach to do here. It's a good approach for an estimate, but we'll get differing results here depending on whether or not a transform is actually occurring. So it opens the possibility that users will get incorrect results.

I think instead what we need to do here is do the bounding box transform only (since that's always done as a worst-case scenario in transformBoundingBox), and then return false to indicate to the caller QgsVectorLayerFeatureIterator that the distance within check needs to be done locally (and accordingly we need to remove the limit from the provider's request).

strk replied:
Sorry I don't fully understand the code, I just tried to improve what I was seeing already done.
What was done was setting mFilter to the reprojected bounding box and later re-setting the request bounding box to that rectangle:

     mRequest.setFilterRect( mFilterRect );

Maybe this was the problem: modifying the request in the first place ?

I'm not clear what doing things "locally" means (by QGIS instead of by provider?).

If you see what happens with the testcase included in this PR, my patch makes the filtering happen correctly on the provider's side (PostgreSQL case) and keeps the LIMIT.

nyalldawson commented:
Superseded by https://github.com/qgis/QGIS/pull/46429

strk commented:
I'm reopening this PR to add a test for the mismatching CRS case which @nyalldawson wrote a commit to support in his #46429 - the aim here is to see it failing _unless_ commit b1f8178a0549f27cb831106dd5fd85d3a4e15c70 from his branch is added

strk commented:
GitHub did not let me reopen this, so I'm continuing in #46971
